### PR TITLE
[DNM] Light Replacer Snazzification

### LINKED ISSUE #4972: (WEB REPORT BY: Anonymous REMOTE: ) Light Replacer does not implement high-efficiency bulbs
>Revision (Should be above if you're viewing this from ingame!)
1767fa9a160eab35f2b32313ec05cf0b9956517e 

>General description of the issue
When an empty light replacer is filled with high efficiency bulbs, the replacer still inserts regular tubes into light fixtures
though its probably just the way the replacer works, it is quite annoying and weird to have regular bulbs be inserted after all that

>What you expected to happen
the high efficiency bulbs to be inserted
>What actually happened
a normal bulb

>Steps to reproduce if possible
fill empty replacer with HE bulbs, replace an empty light fixture

### LINKED ISSUE #4988: Light replacer bug
The light replacer converts any High Efficiency lights into normal lights when you put them in it.

### PR DESCRIPTION
**Pull request by Exxion**
I started working on this yesterday. It's almost finished, but not quite.
To do:
- [x] Change the one in the janitor's locker to loaded/mixed
- [ ] Make borg light replacers... work
- ~~Possibly make advanced variant with more customizability in where it puts removed lights (ex: option to only put HE lights in supply and put normal lights in waste whether they work or not)~~ Nah I'll do that later (if ever)
- [x] Think of more things for this list
- [x] Fix silly description error and huge space in interface window
- [ ] Changelog
- [x] Re-add manufacturing of bulbs from glass
- [x] `target.rigged = emagged || best_light.rigged`

This PR greatly alters the light replacer. Rather than magically transmuting all lights and glass sheets into whatever light it happens to be used on and dropping the old one on the ground, the light replacer now draws lights from an inventory and places the ones taken from fixtures into an inventory.
- Light replacer now holds two boxes
  - Only accepts /obj/item/weapon/storage/box/lights and its subtypes for these boxes
  - One box is its light supply
    - Lights inserted into fixtures are drawn from this supply
    - Chooses the "best" light in its supply box that is compatible with the fixture it is used on
      - At the moment, it just prioritizes HE lights over regular ones
    - If it has a light that is "better" than the current one in the fixture, it will replace the one in the fixture, regardless of whether or not the one in the fixture works
      - If it does work, it is placed into the supply box
    - Interface popup (accessed by clicking the light replacer in your active hand) says how many of each type of light the supply box has
  - The other box is its waste container
    - Broken or burnt-out lights inserted into the light replacer go in here
    - Interface readout says how full this is, but not what types of lights it contains, as they should all be broken
  - Interface popup allows either box to be ejected
  - Boxes can, of course, be inserted
    - If you attempt to insert a box, it first tries to use it as the supply, then tries to use it as the waste container if it already has a supply container. Obviously fails if it has both.
  - Lights can be inserted the same way as before, too, though it's generally much slower than just replacing the supply/waste box
    - Working ones inserted this way go into the supply box, broken/burnt-out ones go into the waste box

Webm demo: https://www.dropbox.com/s/8s2hployrjt155y/2015-06-01_16-20-05.webm?dl=0

Fixes #4972 
Fixes #4988

### CODE REVIEW

**Comment by PJB3005** [comment]
snazzy

**Comment by 9600bauds** [comment]
Literally the only reason why I have ever bothered with the lights replacer is the ability to load it with glass sheets. Were I not be able to do that I would immedately disposals it, because there is nothing worse in life than having to ORDER replacement lights or print them painstakingly from an Autolathe. Even with glass sheets as ammo, it's still more work to replace lights than break them, which is why we still get people breaking all the lights they can.

**Comment by Exxion** [comment]
I may add the glass sheets thing back in. Not sure.
But this does make it more convenient to use because it automatically picks the old light back up and sticks it in its waste container (unless it works in which case it sticks it in its supply container)

But the janitor starts with several boxes of lights and there are more lying around in emergency storage rooms. And since the thing is loaded with boxes, refilling it is as simple as removing the old box and inserting a new one. No more taking each light out of the box and putting them in one-by-one.

**Comment by 9600bauds** [comment]
Believe me, you go through boxes fast.

**Comment by Exxion** [comment]
I think the janitor has 84 bulbs in total across all the boxes in the closet, and 105 if you count the box this starts with (or will once I change the one in the closet to start with a box)
That said, I do need to do something to let it work for borgs. Maybe I'll let it process glass sheets into lights.

**Comment by 9600bauds** [comment]
Janitor's not the only one to replace lights though, Science can make more lights replacers, I often make 5 and chuck them out the window since the janitor is more often than not dead or does not give a shit.

**Comment by Exxion** [comment]
The problem is that, since it now uses lights from storage rather than creating them on use, it would somehow have to be decided which type of light to make out of glass. I guess I could add an option to its popup to select the type of light to make.

In most cases, it would still be faster just to switch out the supply box, but if there are no more light boxes around, that would pose a bit of a problem. That would also be a problem if you are a borg.

I'm also not sure how to deal with the waste container of a borg's light replacer. I guess I could just give it like a billion storage slots and make it impossible to remove, but that would be fucking stupid and I would only do that if I could think of absolutely nothing else.

**Comment by CptWad** [comment]
Does this still have the emag properties? This is important the best part about traitor janitor is glass shard cannons and emagged light replacer bomb set ups.

**Comment by Exxion** [comment]
It should. I didn't change any of the code governing its emag effects. I haven't tested emagging it yet, but I see no reason it wouldn't work.

**Comment by Exxion** [comment]
Been busy for the past few days and don't feel well today, but I am still working on this

**Comment by 9600bauds** [comment]
No rush

**Comment by Exxion** [comment]
Conflicts should be trivial to fix but I'm STILL BUSY
I *should* be able to finish this up on Wednesday

**Comment by Kurfursten** [comment]
I like it, but I definitely agree we need to be able to insert glass to generate bulbs. A simple pop-up asking if you want tube or bulb would suffice. Maybe plasmaglass would make HE bulbs?

**Comment by Exxion** [comment]
Holy shit I rebased without fucking everything up

**Comment by Exxion** [comment]
I changed m_amt and g_amt for some types of lights. They're still low enough that the differences should be trivial for most circumstances, though. The most expensive light is still less than a tenth of a sheet of glass and less than a fiftieth of a sheet of metal.

**Comment by Exxion** [comment]
It can now make lights out of glass again, but it uses much more glass than an autolathe does to make the same light and lights made this way start out with a moderately high switchcount, meaning they're more likely to burn out sooner than other lights.

Borg light replacers hold more glass than regular ones, use less to fabricate lights, and make lights that last just as long as any others. I need to edit the borg light replacer's menu before it'll actually be ready. Also I have to test the light fabrication stuff.

File: code/game/objects/items/devices/lightreplacer.dm [hunk 1]
```diff
@@ -114,72 +156,102 @@
...
+	if(waste)
+		dat += {"<br><br><br><h3>Waste Container:</h3>
+	
+		<b>Filled: </b>[waste.contents.len]/[waste.storage_slots]<br>
+		<b><a href='?src=\ref[src];eject=waste'>Eject Waste Container</a></b>
+		"}
+	else
+		dat += "<br><br><br><h3>No waste container inserted</h3>"
+
+	user << browse(dat, "window=lightreplacer")
```
Comment (PJB3005) [inline code/game/objects/items/devices/lightreplacer.dm]: Oi, browser datums

Comment (Exxion) [inline code/game/objects/items/devices/lightreplacer.dm]: What?
Sorry, not sure what you mean.

Comment (PJB3005) [inline code/game/objects/items/devices/lightreplacer.dm]: https://github.com/d3athrow/vgstation13/blob/Bleeding-Edge/code/datums/browser.dm

They're a basic framework for interfaces, and also has some default CSS to make them look not shit.

Comment (Exxion) [inline code/game/objects/items/devices/lightreplacer.dm]: OK, I'll work on changing that when I get a chance.

**Comment by ComicIronic** [comment]
Happening when?

**Comment by Exxion** [comment]
Sorry. I've been busy. Should have it done by middle of next week at the latest. All that needs to be done:
- Fix conflicts (should be pretty much trivial; just material datums stuff)
- Clean one thing up that I later learned how to do better
- Remake menu with browser datums
- Make alternate menu for borg light replacers
- Changelog

**Comment by ComicIronic** [comment]
Consider finishing PRs before opening them.

**Comment by Exxion** [comment]
Sorry. I thought I wouldn't be busy after that, but I was. I'll reopen it when it's done.

**Comment by Exxion** [comment]
Someone should force me to replace all the lights on the station with HE lights again next time I play Janitor and have free time to finish this
That'll ensure I actually _do_ finish it
This makes it orders of magnitude easier

I still only have to fix some minor conflicts, finish the menus, and fix a couple bugs
And I'll have the time sometime this week
So all I need is the motivation

**Pull request closed without merging**